fabtests/unit/ep_test.c

```diff
@@ -0,0 +1,444 @@
+	unsigned int num_actions = num_eps;
+	int *open_res_arr;
+
+	if (seed == -1)
+		seed = time(NULL);
+	srand(seed);
+
+	open_res_arr = calloc(max_eps, sizeof(*open_res_arr));
+	if (!open_res_arr) {
+		sprintf(err_buf, "Out of memory");
+		return FAIL;
+	}
+
+	opened_count = closed_count = 0;
+	for (i = 0; i < num_actions; ++i) {
+		int idx = rand() % max_eps;
+
+		if (!open_res_arr[idx]) {
+			++opened_count;
+			ret = open_ep(idx);
+			if (ret) {
+				printf("seed = %d\n", seed);
+				goto out;
+			}
+			open_res_arr[idx] = 1;
+		} else {
+			++closed_count;
+			close_ep(idx);
+			open_res_arr[idx] = 0;
+		}
+	}
+	testret = PASS;
+out:
+	for (i = 0; i < max_eps; ++i) {
+		close_ep(i);
+		open_res_arr[i] = 0;
+	}
+	free(open_res_arr);
+	return TEST_RET_VAL(ret, testret);
+}
+
+static int ep_check_flags(void)
+{
+	int ret, testret = FAIL;
+	int ep_index = 1;
+	uint64_t flags;
+	struct fid_ep *ep;
+
+	ret = open_ep(ep_index);
+	if (ret)
+		goto out;
+
+	ep = ep_arr[ep_index];
+	flags = FI_TRANSMIT | FI_TRANSMIT_COMPLETE;
+	ret = fi_control(&ep->fid, FI_SETOPSFLAG, &flags);
+	if (ret) {
+		printf("set tx flag %" PRIu64 " failed!\n", flags);
+		goto out;
+	}
+
+	ret = fi_control(&ep->fid, FI_GETOPSFLAG, &flags);
+	if (ret || flags != FI_TRANSMIT_COMPLETE) {
+		printf("get result of set tx flag %" PRIu64 " failed!\n", flags);
+		goto out;
+	}
+
+	flags = FI_RECV | FI_COMMIT_COMPLETE;
+	ret = fi_control(&ep->fid, FI_SETOPSFLAG, &flags);
+	if (ret) {
+		printf("set rx flag %" PRIu64 " failed!\n", flags);
+		goto out;
+	}
+
+	ret = fi_control(&ep->fid, FI_GETOPSFLAG, &flags);
+	if (ret || flags != FI_COMMIT_COMPLETE) {
+		printf("get result of rx flag %" PRIu64 " failed!\n", flags);
+		goto out;
+	}
+
+	testret = PASS;
+out:
+	close_ep(ep_index);
+	return TEST_RET_VAL(ret, testret);
+}
+
+struct test_entry test_array[] = {
+	TEST_ENTRY(ep_open_close, "Test open and close EPs"),
+	TEST_ENTRY(ep_open_max_simultaneously, "Test opening maximum number of EPs at the same time"),
+	TEST_ENTRY(ep_rand_access, "Test opening and closing EPs randomly"),
+	TEST_ENTRY(ep_open_sizes, "Test opening and closing EPs with different context sizes"),
+	TEST_ENTRY(ep_check_flags, "Test get and set EP operation flags"),
+	{ NULL, "" }
+};
+
+static void usage(void)
+{
+	ft_unit_usage("ep_test", "Unit test for end point (EP)");
+	FT_PRINT_OPTS_USAGE("-n <num>", "number of EPs to open");
+	FT_PRINT_OPTS_USAGE("-s <seed>", "seeds the random number generator");
+}
+
+int main(int argc, char **argv)
+{
+	int op, ret;
+	int failed;
+
+	hints = fi_allocinfo();
+	if (!hints)
+		return EXIT_FAILURE;
+
+	opts = INIT_OPTS;
+	opts.options |= FT_OPT_SIZE;
+	opts.options |= FT_OPT_SKIP_REG_MR;
+	hints->ep_attr->type = FI_EP_RDM;
+	num_eps = 50;
+
+	while ((op = getopt(argc, argv, INFO_OPTS "hn:s:")) != -1) {
+		switch (op) {
+		default:
+			ft_parseinfo(op, optarg, hints, &opts);
+			break;
+		case 'n':
+			num_eps = atoi(optarg);
+			break;
+		case 's':
```

Comment:
Can we pick an unused option, since 's' is used by other tests for another purpose?